### PR TITLE
Added deepClone functionality to trie

### PR DESCRIPTION
**Pull request by iulianpascalau**
Added deepClone functionality to trie
Removed old, unused files from data/mock package

### CODE REVIEW

File: data/trie/branchNode_test.go [hunk 1]
```diff
@@ -791,6 +794,101 @@ func newEmptyTrie() data.Trie {
...
+		if bn.children[i] == nil {
+			childsString += "<nil>"
+			continue
+		}
+
+		childsString += fmt.Sprintf("%p", bn.children[i])
+	}
+
+	str := fmt.Sprintf(`extension node:
+   encoded child: %s
```
Comment (sasurobert) [inline data/trie/branchNode_test.go]: indent these lines

File: data/trie/extensionNode_test.go [hunk 2]
```diff
@@ -591,3 +594,109 @@ func TestExtensionNode_isEmptyOrNil(t *testing.T) {
...
+			getExtensionNodeContents(expected),
+			getExtensionNodeContents(actual),
+		)
+	}
+	assert.False(t, expected == actual)
+}
+
+func getExtensionNodeContents(en *extensionNode) string {
+	str := fmt.Sprintf(`extension node:
+   key: %s
```
Comment (sasurobert) [inline data/trie/extensionNode_test.go]: indent these lines

File: data/trie/leafNode_test.go [hunk 3]
```diff
@@ -378,3 +382,96 @@ func TestLeafNode_isEmptyOrNil(t *testing.T) {
...
+	maxChars := 32
+	buff := make([]byte, maxChars)
+	_, _ = rand.Reader.Read(buff)
+
+	return buff
+}
+
+func getLeafNodeContents(lf *leafNode) string {
+	str := fmt.Sprintf(`leaf node:
+   key: %s
```
Comment (sasurobert) [inline data/trie/leafNode_test.go]: indent these lines

File: data/trie/branchNode.go [hunk 4]
```diff
@@ -496,3 +496,36 @@ func (bn *branchNode) print(writer io.Writer, index int) {
 		child.print(writer, index+len(str)-1+len(str2))
 	}
 }
+
+func (bn *branchNode) deepClone() node {
```
Comment (BeniaminDrasovean) [inline data/trie/branchNode.go]: You should check if bn.isEmptyOrNil(), as all the other methods do.

File: data/trie/extensionNode.go [hunk 5]
```diff
@@ -378,3 +378,30 @@ func (en *extensionNode) print(writer io.Writer, index int) {
 	_, _ = fmt.Fprint(writer, str)
 	en.child.print(writer, index+len(str))
 }
+
+func (en *extensionNode) deepClone() node {
```
Comment (BeniaminDrasovean) [inline data/trie/extensionNode.go]: You should check if en.isEmptyOrNil(), as all the other methods do.

File: data/trie/leafNode.go [hunk 6]
```diff
@@ -241,3 +241,26 @@ func (ln *leafNode) print(writer io.Writer, index int) {
 
 	_, _ = fmt.Fprintf(writer, "L:(%s - %s)\n", key, val)
 }
+
+func (ln *leafNode) deepClone() node {
```
Comment (BeniaminDrasovean) [inline data/trie/leafNode.go]: You should check if ln.isEmptyOrNil(), as all the other methods do.

File: data/trie/patriciaMerkleTrie_test.go [hunk 7]
```diff
@@ -327,6 +327,21 @@ func emptyTrie() data.Trie {
...
 }
 
+func TestPatriciaMerkleTrie_DeepCloneShouldWork(t *testing.T) {
+	tr := initTrie()
+
+	_ = tr.Update([]byte("doee"), []byte("value of doee"))
+	_ = tr.Update([]byte("doeee"), []byte("value of doeee"))
+
+	trie2, err := tr.DeepClone()
+
```
Comment (BeniaminDrasovean) [inline data/trie/patriciaMerkleTrie_test.go]: delete empty line

File: data/trie/patriciaMerkleTrie_test.go [hunk 8]
```diff
@@ -327,6 +327,21 @@ func emptyTrie() data.Trie {
...
+	assert.Nil(t, err)
+
+	assert.Equal(t, tr, trie2)
+	assert.False(t, tr == trie2)
+	assert.Equal(t, tr.String(), trie2.String())
+}
```
Comment (BeniaminDrasovean) [inline data/trie/patriciaMerkleTrie_test.go]: As another method to see if the clone was done correctly, you could compute the rootHash of both tries, and then compare the hashes.

File: data/trie/patriciaMerkleTrie_test.go [hunk 9]
```diff
@@ -538,19 +553,40 @@ func BenchmarkPatriciaMerkleTrie_RootHashAfterChanging30000Nodes(b *testing.B) {
...
+
+	nrValuesInTrie := 10000
+	values := make([][]byte, nrValuesInTrie)
+
+	for i := 0; i < nrValuesInTrie; i++ {
+		key := hsh.Compute(strconv.Itoa(i))
+		value := append(key, []byte(strconv.Itoa(i))...)
+
+		_ = tr.Update(key, value)
+		values[i] = key
```
Comment (BeniaminDrasovean) [inline data/trie/patriciaMerkleTrie_test.go]: I think you can remove this line, and also line 577.

File: data/trie/branchNode_test.go (hunk 1 above)
Comment (iulianpascalau) [reply data/trie/branchNode_test.go]: done, added 2 more tabs for each line

File: data/trie/leafNode_test.go (hunk 3 above)
Comment (iulianpascalau) [reply data/trie/leafNode_test.go]: done, added 2 more tabs for each line

File: data/trie/branchNode.go (hunk 4 above)
Comment (iulianpascalau) [reply data/trie/branchNode.go]: ok, will add checks. The rest of our project lacks these kind of tests.

File: data/trie/extensionNode.go (hunk 5 above)
Comment (iulianpascalau) [reply data/trie/extensionNode.go]: Checked only for nil

File: data/trie/leafNode.go (hunk 6 above)
Comment (iulianpascalau) [reply data/trie/leafNode.go]: Checked only for nil

File: data/trie/patriciaMerkleTrie_test.go (hunk 7 above)
Comment (iulianpascalau) [reply data/trie/patriciaMerkleTrie_test.go]: done

File: data/trie/patriciaMerkleTrie_test.go (hunk 8 above)
Comment (iulianpascalau) [reply data/trie/patriciaMerkleTrie_test.go]: I think `assert.Equal(t, tr, trie2)` will suffice. However, I have added the assert as a supplementary check against RootHash algorithm

File: data/trie/patriciaMerkleTrie_test.go (hunk 9 above)
Comment (iulianpascalau) [reply data/trie/patriciaMerkleTrie_test.go]: done

File: data/trie/patriciaMerkleTrie_test.go [hunk 10]
```diff
@@ -327,6 +327,25 @@ func emptyTrie() data.Trie {
...
+	_ = tr.Update([]byte("doeee"), []byte("value of doeee"))
+
+	trie2, err := tr.DeepClone()
+	assert.Nil(t, err)
+
+	assert.Equal(t, tr, trie2)
+	assert.False(t, tr == trie2)
+	assert.Equal(t, tr.String(), trie2.String())
+	originalRoot, _ := tr.Root()
+	clonedTrie, _ := tr.Root()
```
Comment (BeniaminDrasovean) [inline data/trie/patriciaMerkleTrie_test.go]: trie2.Root() instead of tr.Root()

Comment (iulianpascalau) [reply data/trie/patriciaMerkleTrie_test.go]: Right, fixed!

**Pull request merged**